server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/processors/FunctionDeclarationProcessor.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2024 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Broadcom, Inc. - initial API and implementation
+ *
+ */
+package org.eclipse.lsp.cobol.core.engine.processors;
+
+import org.eclipse.lsp.cobol.common.error.ErrorSeverity;
+import org.eclipse.lsp.cobol.common.error.ErrorSource;
+import org.eclipse.lsp.cobol.common.error.SyntaxError;
+import org.eclipse.lsp.cobol.common.message.MessageTemplate;
+import org.eclipse.lsp.cobol.common.model.tree.FunctionDeclaration;
+import org.eclipse.lsp.cobol.common.model.tree.FunctionReference;
+import org.eclipse.lsp.cobol.common.processor.ProcessingContext;
+import org.eclipse.lsp.cobol.common.processor.Processor;
+import org.eclipse.lsp.cobol.core.engine.symbols.SymbolAccumulatorService;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/** Processor for FUnctionDeclaration */
+public class FunctionDeclarationProcessor implements Processor<FunctionDeclaration> {
+  private final SymbolAccumulatorService symbolAccumulatorService;
+
+  private static final String WHEN_COMPILED = "WHEN-COMPILED";
+  private static final String ALL = "ALL";
+
+  public FunctionDeclarationProcessor(SymbolAccumulatorService symbolAccumulatorService) {
+    this.symbolAccumulatorService = symbolAccumulatorService;
+  }
+
+  @Override
+  public void accept(FunctionDeclaration functionDeclaration, ProcessingContext processingContext) {
+    updateFunctionInfo(functionDeclaration);
+    if (functionDeclaration.isIntrinsic()) {
+      checkSemanticsForIntrinsicFunctions(functionDeclaration, processingContext);
+    } else if (functionDeclaration.getChildren().size() > 1) {
+      checkSemanticsForUserDefinedFunctions(functionDeclaration, processingContext);
+    }
+  }
+
+  private void updateFunctionInfo(FunctionDeclaration functionDeclaration) {
+    getFunctionReferences(functionDeclaration)
+        .forEach(
+            reference -> {
+              SymbolAccumulatorService.FunctionInfo functionInfo =
+                  symbolAccumulatorService.getFunctionReference(reference.getName().toUpperCase());
+              if (functionInfo != null) functionInfo.setDeclared(true);
```

Comment:
This breaks in the multi-program case, doesn't it? Basically, once the function is declared in the repository, it will be declared in all programs that follow.